### PR TITLE
Move user dim logic into new UserDimContainer

### LINKED ISSUE #4204: First game with autoplay on any map removes background dim completely
How to reproduce: just start any map with autoplay (by adding a mod or pressing Ctrl+Enter - no matter). After few seconds background dim will be removed. Btw dim Bindable isn't changing.
Example (thanks @KoziLord): https://streamable.com/3fhjl

### PR DESCRIPTION
**Pull request by nyquillerium**
Closes #4204.

- [x] Depends on #4296

### CODE REVIEW

File: osu.Game/Screens/Play/Player.cs [hunk 1]
```diff
@@ -395,6 +397,8 @@ public override bool OnExiting(IScreen next)
             if (LoadedBeatmapSuccessfully)
                 pauseContainer?.Pause();
 
+            Background.UpdateDim.Value = false;
```
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs]: why does this use bindable but above uses a method?

File: osu.Game.Tests/Visual/TestCaseBackgroundScreenBeatmap.cs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+using osu.Game.Graphics;
+using osu.Game.Rulesets;
+using osu.Game.Screens;
+using osu.Game.Screens.Backgrounds;
+using osu.Game.Screens.Play;
+
+namespace osu.Game.Tests.Visual
+{
+    [TestFixture]
+    public class TestCaseBackgroundScreenBeatmap : TestCasePlayer
```
Comment (peppy) [inline osu.Game.Tests/Visual/TestCaseBackgroundScreenBeatmap.cs]: this test plays through all four rulesets unnecessarily.

File: osu.Game.Tests/Visual/TestCaseBackgroundScreenBeatmap.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            AddStep("Test User Dimming", () => ((DimAccessiblePlayer)Player).DisableScreenDim());
+            AddWaitStep(5, "Wait for dim");
+            AddAssert("Check screen dim", () => ((DimAccessiblePlayer)Player).AssertUndimmed());
+        }
+
+        protected override Player CreatePlayer(Ruleset ruleset) => new DimAccessiblePlayer();
+
+        private class DimAccessiblePlayer : Player
+        {
+            protected override BackgroundScreen CreateBackground() => new FadeAccessibleBackground();
```
Comment (peppy) [inline osu.Game.Tests/Visual/TestCaseBackgroundScreenBeatmap.cs]: newline between all methods.

File: osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs [hunk 4]
```diff
@@ -2,18 +2,34 @@
...
 namespace osu.Game.Screens.Backgrounds
 {
     public class BackgroundScreenBeatmap : BlurrableBackgroundScreen
     {
         private WorkingBeatmap beatmap;
+        protected Bindable<double> DimLevel;
+        public Bindable<bool> UpdateDim;
 
-        public WorkingBeatmap Beatmap
+        protected float BackgroundOpacity => 1 - (float)DimLevel;
```
Comment (peppy) [inline osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs]: just mode this inline. there's only one usage and following the logic right now feels convoluted.

File: osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs [hunk 5]
```diff
@@ -35,18 +52,51 @@ public WorkingBeatmap Beatmap
...
+            UpdateDim.Value = false;
+            return base.OnExiting(last);
+        }
+
+        private void updateBackgroundDim()
+        {
+            if (UpdateDim)
+                FadeContainer?.FadeColour(OsuColour.Gray(BackgroundOpacity), 800, Easing.OutQuint);
+            else
+                FadeContainer?.FadeColour(OsuColour.Gray(1.0f), 800, Easing.InQuint);
```
Comment (peppy) [inline osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs]: The easing logic was changed inadvertently?

```suggestion
                FadeContainer?.FadeColour(OsuColour.Gray(1.0f), 800, Easing.OutQuint);
```

Comment (peppy) [reply osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs]:
```suggestion
                FadeContainer?.FadeColour(Color4.White, 800, Easing.InQuint);
```

File: osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs [hunk 6]
```diff
@@ -35,18 +52,51 @@ public WorkingBeatmap Beatmap
...
+        }
+
         public BackgroundScreenBeatmap(WorkingBeatmap beatmap = null)
         {
             Beatmap = beatmap;
+            UpdateDim = new Bindable<bool>();
+        }
+
+        public void EnableUserDim()
```
Comment (peppy) [inline osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs]: why not just use the bindable directly? as i mentioned in my previous comment: is there any reason to have methods for this?

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Play/Player.cs [hunk 7]
```diff
@@ -61,6 +61,7 @@ public class Player : ScreenWithBeatmapBackground, IProvideCursor
 
         public CursorContainer Cursor => RulesetContainer.Cursor;
         public bool ProvidingUserCursor => RulesetContainer?.Cursor != null && !RulesetContainer.HasReplayLoaded.Value;
+        protected float BackgroundOpacity => 1 - (float)DimLevel;
```
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs]: never used?

File: osu.Game/Screens/Play/Player.cs [hunk 8]
```diff
@@ -387,13 +390,15 @@ public override bool OnExiting(IScreen next)
...
                 applyRateFromMods();
-
                 fadeOut();
                 return base.OnExiting(next);
             }
 
             if (LoadedBeatmapSuccessfully)
+            {
                 pauseContainer?.Pause();
+                return true;
```
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs]: no point in this?

File: osu.Game/Screens/Play/ScreenWithBeatmapBackground.cs [hunk 9]
```diff
@@ -40,6 +37,7 @@ private void load(OsuConfigManager config)
 
         public override void OnEntering(IScreen last)
         {
+            // We need to update on dim here because player still needs to know if it needs to dim the storyboard
```
Comment (peppy) [inline osu.Game/Screens/Play/ScreenWithBeatmapBackground.cs]: i don't really understand this comment

File: osu.Game/Screens/Play/PlayerLoader.cs [hunk 10]
```diff
@@ -243,6 +243,9 @@ public override bool OnExiting(IScreen next)
             this.FadeOut(150);
             cancelLoad();
 
+            if (Background != null)
+                Background.UpdateDim.Value = false;
```
Comment (peppy) [inline osu.Game/Screens/Play/PlayerLoader.cs]: a single class shouldn't set dim to disabled when it never enabled it. if this is because player enabled updatedim, then shouldn't player disable it?

there's two ways you can look at it:

- If player enables in enter, it should disable in exit. (apply to every usage)
- Always set enabled or disabled in OnEntering and OnResuming in every class that has a beatmap background. (no screen undoes its applications, but it's always correct because every screen applies).

File: osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs [hunk 11]
```diff
@@ -35,18 +52,43 @@ public WorkingBeatmap Beatmap
...
+
+        private void updateBackgroundDim()
+        {
+            FadeContainer?.FadeColour(UpdateDim ? OsuColour.Gray(1 - (float)DimLevel) : Color4.White, 800, Easing.OutQuint);
+        }
+
         public BackgroundScreenBeatmap(WorkingBeatmap beatmap = null)
         {
             Beatmap = beatmap;
+            UpdateDim = new Bindable<bool>();
```
Comment (peppy) [inline osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs]: move to field initialiser

File: osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs [hunk 12]
```diff
@@ -35,18 +52,43 @@ public WorkingBeatmap Beatmap
...
         }
 
+        public override void OnEntering(IScreen last)
+        {
+            base.OnEntering(last);
+            DimLevel.ValueChanged += _ => updateBackgroundDim();
+            UpdateDim.ValueChanged += _ => updateBackgroundDim();
+            updateBackgroundDim();
+        }
+        public override void OnResuming(IScreen last)
```
Comment (peppy) [inline osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs]: missing newline

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Play/Player.cs (hunk 7 above)
Comment (nyquillerium) [reply osu.Game/Screens/Play/Player.cs]: It is being used by the storyboard still, because storyboardContainer is still a part of Player

File: osu.Game/Screens/Play/PlayerLoader.cs (hunk 10 above)
Comment (nyquillerium) [reply osu.Game/Screens/Play/PlayerLoader.cs]: this is actually some left over code from when I was testing a bug, shouldn't be necessary here.

File: osu.Game/Screens/Play/Player.cs [hunk 13]
```diff
@@ -345,6 +345,11 @@ public override void OnEntering(IScreen last)
                 .Delay(250)
                 .FadeIn(250);
 
+            ShowStoryboard.ValueChanged += _ => UpdateBackgroundElements();
```
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs]: is this actually required? i mean the part where it's calling the base logic. it doesn't feel right since i'm pretty sure you already have bindables taking care of this flow.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Play/Player.cs (hunk 13 above)
Comment (nyquillerium) [reply osu.Game/Screens/Play/Player.cs]: Moving the bindable update triggered storyboard initialization here to resolve this.

File: osu.Game/Graphics/Containers/UserDimContainer.cs [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+        {
+            DimLevel = config.GetBindable<double>(OsuSetting.DimLevel);
+            ShowStoryboard = config.GetBindable<bool>(OsuSetting.ShowStoryboard);
+            EnableUserDim.ValueChanged  += _ => updateBackgroundDim();
+            DimLevel.ValueChanged  += _ => updateBackgroundDim();
+            ShowStoryboard.ValueChanged += _ => updateBackgroundDim();
+            StoryboardReplacesBackground.ValueChanged += _ => updateBackgroundDim();
+        }
+
+        private void updateBackgroundDim()
```
[the remark this answers is not in the document]
Comment (nyquillerium) [reply osu.Game/Graphics/Containers/UserDimContainer.cs]: It is being called whenever EnableUserDim is changed, which happens for the first time when either Player is entered or PlayerLoader triggers a preview. 

It makes sense for it to be always called on load though, will change.

File: osu.Game/Screens/Play/Player.cs [hunk 15]
```diff
@@ -241,6 +239,8 @@ private void load(AudioManager audio, APIAccess api, OsuConfigManager config)
             if (ShowStoryboard.Value)
                 initializeStoryboard(false);
 
+            storyboardContainer.EnableUserDim.Value = true;
```
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs]: just set in the object initialiser?

File: osu.Game/Screens/Play/Player.cs [hunk 16]
```diff
@@ -402,7 +411,8 @@ private void fadeOut(bool instant = false)
         {
             float fadeOutDuration = instant ? 0 : 250;
             this.FadeOut(fadeOutDuration);
-            Background?.FadeColour(Color4.White, fadeOutDuration, Easing.OutQuint);
+            Background.EnableUserDim.Value = false;
+            storyboardContainer.StoryboardReplacesBackground.Value = false;
```
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs]: any point in this?

File: osu.Game/Screens/Play/Player.cs [hunk 17]
```diff
@@ -345,6 +345,16 @@ public override void OnEntering(IScreen last)
...
 
+            ShowStoryboard.ValueChanged += _ =>
+            {
+                if (ShowStoryboard.Value && storyboard == null)
+                    initializeStoryboard(true);
+            };
+
+            Background.EnableUserDim.Value = true;
+            storyboardContainer.StoryboardReplacesBackground.Value = Beatmap.Value.Storyboard.ReplacesBackground && Beatmap.Value.Storyboard.HasDrawable;
+            storyboardContainer.StoryboardReplacesBackground.BindTo(Background?.StoryboardReplacesBackground);
```
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs]: this is weird binding logic. make a local private bendable rather than binding two other classes' bindables

File: osu.Game/Screens/Play/Player.cs [hunk 18]
```diff
@@ -345,6 +345,16 @@ public override void OnEntering(IScreen last)
                 .Delay(250)
                 .FadeIn(250);
 
+            ShowStoryboard.ValueChanged += _ =>
+            {
+                if (ShowStoryboard.Value && storyboard == null)
```
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs]: use valuechanged arg instead of reference to bindable

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Play/Player.cs (hunk 16 above)
Comment (nyquillerium) [reply osu.Game/Screens/Play/Player.cs]: When exiting player, if player had a storyboard that disables the background, whatever screen it suspends/exits to needs to have a visible background.

I added a new test after realizing this.

Comment (peppy) [reply osu.Game/Screens/Play/Player.cs]: This shouldn't be set on this bindable though, since this container is basically out of scope at this point (see my other comments).

Comment (nyquillerium) [reply osu.Game/Screens/Play/Player.cs]: will set on the local private bindable inside Player then

File: osu.Game/Screens/Play/Player.cs [hunk 19]
```diff
@@ -345,6 +346,18 @@ public override void OnEntering(IScreen last)
...
 
+            ShowStoryboard.ValueChanged += s =>
+            {
+                if (s.NewValue && storyboard == null)
+                    initializeStoryboard(true);
+            };
+
+            Background.EnableUserDim.Value = true;
+
+            storyboardReplacesBackground.BindTo(Background?.StoryboardReplacesBackground);
```
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs]: how can `Background` be null? see line above.

File: osu.Game/Screens/Play/Player.cs [hunk 20]
```diff
@@ -345,6 +346,18 @@ public override void OnEntering(IScreen last)
...
+            storyboardReplacesBackground.BindTo(storyboardContainer.StoryboardReplacesBackground);
```
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs]: bind order is back to front (see `BindTo` xmldoc)

File: osu.Game/Screens/Play/Player.cs [hunk 21]
```diff
@@ -345,6 +346,18 @@ public override void OnEntering(IScreen last)
...
+            storyboardReplacesBackground.Value = Beatmap.Value.Storyboard.ReplacesBackground && Beatmap.Value.Storyboard.HasDrawable;
```
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs]: as above

File: osu.Game/Screens/Play/PlayerLoader.cs [hunk 22]
```diff
@@ -151,23 +151,27 @@ protected override void LogoArriving(OsuLogo logo, bool resuming)
...
-        private VisualSettings visualSettings;
+        protected VisualSettings VisualSettings;
 
         private bool readyForPush => player.LoadState == LoadState.Ready && IsHovered && GetContainingInputManager()?.DraggedDrawable == null;
 
         protected override bool OnHover(HoverEvent e)
         {
             // restore our screen defaults
             InitializeBackgroundElements();
+            if (this.IsCurrentScreen() && (Background?.IsLoaded ?? false))
```
Comment (peppy) [inline osu.Game/Screens/Play/PlayerLoader.cs]: is it actually possible for background to be null here? possibly cause for concern.

File: osu.Game/Screens/Play/PlayerLoader.cs [hunk 23]
```diff
@@ -151,23 +151,27 @@ protected override void LogoArriving(OsuLogo logo, bool resuming)
...
+                Background.EnableUserDim.Value = false;
             return base.OnHover(e);
         }
 
         protected override void OnHoverLost(HoverLostEvent e)
         {
-            if (GetContainingInputManager()?.HoveredDrawables.Contains(visualSettings) == true)
+            if (GetContainingInputManager()?.HoveredDrawables.Contains(VisualSettings) == true)
             {
                 // show user setting preview
```
Comment (peppy) [inline osu.Game/Screens/Play/PlayerLoader.cs]: update comment to mention that `UpdateBackgroundElements` is only being called because blur hasn't been moved to `UserDimContainer` (and maybe link the issue mentioning that it needs to be moved).

File: osu.Game/Screens/Play/PlayerLoader.cs [hunk 24]
```diff
@@ -241,6 +245,9 @@ public override bool OnExiting(IScreen next)
             this.FadeOut(150);
             cancelLoad();
 
+            if (Background != null)
```
Comment (peppy) [inline osu.Game/Screens/Play/PlayerLoader.cs]: Again, a bit confused about the null check here. looks impossible.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Play/Player.cs (hunk 21 above)
Comment (nyquillerium) [reply osu.Game/Screens/Play/Player.cs]: Not quite sure what this is referring to. It's setting the value as a one-off thing, not binding to a bindable.

Comment (peppy) [reply osu.Game/Screens/Play/Player.cs]: already fine, was attached to the second BindTo line before

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
I can't run the tests in visual tests. If they aren't run in order there are random crashes.

Also when run in order they seem completely wrong. Steps like "wait for player to load" are triggering pause menu coming up and other weird stuff.

Likely wasn't even tested properly because they are so hard to follow. The storyboard is invisible, the "song select" is invisible.

Tests are supposed to be visually followable.

**Comment by nyquillerium** [comment]
must have broke in one of the changes I made Sunday, will re-test.

File: osu.Game.Tests/Visual/TestCaseBackgroundScreenBeatmap.cs [hunk 25]
```diff
@@ -0,0 +1,421 @@
...
+            AddAssert("Screen is dimmed", () => songSelect.IsBackgroundDimmed());
+        }
+
+        /// <summary>
+        /// In the case of a user triggering the dim preview the instant player gets loaded, then moving the cursor off of the visual settings:
+        /// The OnHover of PlayerLoader will trigger, which could potentially trigger an undim unless checked for in PlayerLoader.
+        /// We need to check that in this scenario, the dim is still properly applied after entering player.
+        /// </summary>
+        [Test]
+        public void PlayerLoaderTransitionTest()
```
Comment (peppy) [inline osu.Game.Tests/Visual/TestCaseBackgroundScreenBeatmap.cs]: have you checked this is testing what you say it is? for me the timings don't seem to be correct.

**Comment by peppy** [comment]
- Reduce number of unnecessary test steps (screen stack create / wait can go in setup)
- Change bool buttons like EnableUserDim to be toggles to allow for user interaction in visualtests (and add asserts for a second toggle).

File: osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs [hunk 26]
```diff
@@ -2,18 +2,31 @@
...
+        /// <summary>
+        /// Whether or not user dim settings should be applied to this Background.
+        /// </summary>
+        public Bindable<bool> EnableUserDim = new Bindable<bool>();
+
+        public Bindable<bool> StoryboardReplacesBackground = new Bindable<bool>();
+
+        protected UserDimContainer FadeContainer;
+
+        protected virtual UserDimContainer CreateFadeContainer() => new UserDimContainer { RelativeSizeAxes = Axes.Both };
```
Comment (smoogipoo) [inline osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs]: Probably not necessary to expose `FadeContainer` if this also exists

File: osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs [hunk 27]
```diff
@@ -37,15 +50,19 @@ public WorkingBeatmap Beatmap
                         }
 
                         b.Depth = newDepth;
-                        AddInternal(Background = b);
+                        FadeContainer.Add(Background = b);
                         Background.BlurSigma = BlurTarget;
+                        FadeContainer.StoryboardReplacesBackground.BindTo(StoryboardReplacesBackground);
```
Comment (smoogipoo) [inline osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs]: This is going to bind multiple times

File: osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs [hunk 28]
```diff
@@ -37,15 +50,19 @@ public WorkingBeatmap Beatmap
...
                 });
             }
         }
 
         public BackgroundScreenBeatmap(WorkingBeatmap beatmap = null)
         {
+            FadeContainer = CreateFadeContainer();
+            InternalChild = FadeContainer;
+            EnableUserDim.BindTo(FadeContainer.EnableUserDim);
             Beatmap = beatmap;
```
Comment (smoogipoo) [inline osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs]: Setters take precedence over other stuff in ctors. I.e.

```
MyClass:
    private object myObject;
    ctor(object myObject):
        this.myObject = myObject;
        /* other stuff */
```

File: osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs [hunk 29]
```diff
@@ -37,15 +50,19 @@ public WorkingBeatmap Beatmap
...
+                        FadeContainer.StoryboardReplacesBackground.BindTo(StoryboardReplacesBackground);
                     }));
                 });
             }
         }
 
         public BackgroundScreenBeatmap(WorkingBeatmap beatmap = null)
         {
+            FadeContainer = CreateFadeContainer();
+            InternalChild = FadeContainer;
```
Comment (smoogipoo) [inline osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs]: Same thing here, this line could just be `InternalChild = FadeContainer = CreateFadeContainer()`.

File: osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs [hunk 30]
```diff
@@ -37,15 +50,19 @@ public WorkingBeatmap Beatmap
...
+            EnableUserDim.BindTo(FadeContainer.EnableUserDim);
```
Comment (smoogipoo) [inline osu.Game/Screens/Backgrounds/BackgroundScreenBeatmap.cs]: In this case it doesn't matter, but I'd prefer this to be written as `FadeContainer.EnableUserDim.BindTo(EnableUserDim)`.

In the general sense, when `.BindTo(other)` happens, the values from `other` are instantly copied into the bindable, so this would mean the value from `FadeContainer.EnableUserDim` is coped into `EnableUserDim`.

File: osu.Game/Screens/Play/Player.cs [hunk 31]
```diff
@@ -87,7 +87,14 @@ public class Player : ScreenWithBeatmapBackground, IProvideCursor
         private FailOverlay failOverlay;
 
         private DrawableStoryboard storyboard;
-        private Container storyboardContainer;
+        protected UserDimContainer StoryboardContainer;
```
Comment (smoogipoo) [inline osu.Game/Screens/Play/Player.cs]:
```suggestion
        protected UserDimContainer StoryboardContainer { get; private set; }
```

Along with all other protected members in this class.

File: osu.Game/Screens/Play/Player.cs [hunk 32]
```diff
@@ -346,6 +349,18 @@ public override void OnEntering(IScreen last)
...
 
+            ShowStoryboard.ValueChanged += s =>
+            {
+                if (s.NewValue && storyboard == null)
+                    initializeStoryboard(true);
+            };
+
+            Background.EnableUserDim.Value = true;
+
+            storyboardReplacesBackground.BindTo(Background.StoryboardReplacesBackground);
```
Comment (smoogipoo) [inline osu.Game/Screens/Play/Player.cs]: Same thing here, but it matters more here, the `.BindTo()` direction should be inversed.

File: osu.Game/Screens/Play/Player.cs [hunk 33]
```diff
@@ -346,6 +349,18 @@ public override void OnEntering(IScreen last)
...
+            StoryboardContainer.StoryboardReplacesBackground.BindTo(Background.StoryboardReplacesBackground);
```
Comment (smoogipoo) [inline osu.Game/Screens/Play/Player.cs]:
```suggestion
            StoryboardContainer.StoryboardReplacesBackground.BindTo(storyboardReplacesBackground);
```

**Review by smoogipoo — DISMISSED** [review]

**Review by peppy — APPROVED** [review]
